src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -1438,6 +1470,16 @@ void GcsActorManager::Initialize(const GcsInitData &gcs_init_data) {
     }
   }
   if (!dead_actors.empty()) {
+    for (const auto &actor_id : dead_actors) {
+      auto *actor = GetActor(actor_id);
+      RAY_CHECK(actor != nullptr);
+      RAY_LOG(INFO).WithField(kLogKeyActorID, actor_id)
+          << "Destroying actor because it should be dead when GCS restarts.";
+      if (actor->GetState() != rpc::ActorTableData::DEAD) {
+        // destroy the actor
+        DestroyActor(actor_id, GenActorOutOfScopeCause(actor));
+      }
```

Comment:
I think this should be moved to be under line 1464 `else clause` because we blindly add the actor to destroyed_actors_  without checking whether the state is dead or not.

I think we should do

```
else {
  auto actor = std::make_shared<GcsActor>(actor_table_data, actor_state_counter_);
  if (actor->GetState() == rpc::ActorTableData::DEAD) {
      AddDestroyedActorToCache()
  } else {
     DestroyActor();
  }
 dead_actors.push_back(actor_id);
}
```

Reply:
DestroyActor requires it first be registered. Redoing a bit to first register all non-DEAD actors, then kill should-be-dead actors within.